pmem: make pmem_is_pmem true for Device DAX only

krzycz opened the pull request:
RFC
This is the minimal change required to support new behavior of *pmem_is_pmem* on Linux.
The original plan was that Linux and Windows implementations would share the same code and data structure for tracking all the file mappings. However, to minimize the changes, I decided to keep the Linux and Windows implementations separate for now, so the Windows version of NVML is not affected by this patch.
The overview of the changes:
- on Linux *pmem_is_pmem* would now return true only for Device DAX mappings (unless overriden with PMEM_IS_PMEM_FORCE)
- only PMEM (dev dax) regions are tracked (if region is not on the list, it's not PMEM)
- only the mappings created with *pmem_map_file* are tracked (if the user would query a direct pointer to the object allocated from pmemobj pool, pmem_is_pmem will return false) [file an issue]
- adjacent regions are not merged [to be optimized in the future]
- multiple entries for the same or overlapping region are not detected/handled properly, but it should never happen in case of *pmem_map_file* (no MAP_FIXED) [to be fixed in the future]
- *libpmem* would depend on *libpthread* again!
- the internal data structures holding information about all the file mappings is allocated with malloc() [file an issue - new API must be added]

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/pmem/nvml/1675)
<!-- Reviewable:end -->

andyrudoff commented on doc/libpmem.3.md:
```diff
@@ -164,6 +164,11 @@ appropriate for flushing changes to persistence. Calling
 **pmem_is_pmem**() each time changes are flushed to persistence will
 not perform well.
 
+>NOTE:
```
How about adding an additional sentence to this note to give people a little hope.  "In the future, ss file systems become available that support flushing with **pmem_persist**(), **pmem_is_pmem**() will return true as appropriate."

Also "directly from DAX Device" should be "directly from Device DAX" since that's how the ndctl man page refers to it.

andyrudoff commented on src/libpmem/pmem.c:
```diff
@@ -681,6 +682,13 @@ pmem_map_file(const char *path, size_t len, int flags, mode_t mode,
 
+#ifndef _WIN32
+	/* XXX only Device DAX regions (PMEM) are tracked so far */
+	if (dax && util_range_track(addr, len) != 0) {
```
Very picky comment: can we rename the "dax" flag "device_dax" in this function?

krzycz commented:
Review status: 0 of 88 files reviewed at latest revision, 2 unresolved discussions.

---

*[doc/libpmem.3.md, line 167 at r1](https://reviewable.io:443/reviews/pmem/nvml/1675#-KcxRRbkxmPXMYcUnl1v-r1-167:-Kd-cma-Xv2OQVH3PY9Y:b-896fix) ([raw file](https://github.com/pmem/nvml/blob/f5d50258e779518f79a803873133a8b999313143/doc/libpmem.3.md#L167)):*
<details><summary><i>Previously, andyrudoff (Andy Rudoff) wrote…</i></summary><blockquote>

How about adding an additional sentence to this note to give people a little hope.  "In the future, ss file systems become available that support flushing with **pmem_persist**(), **pmem_is_pmem**() will return true as appropriate."

Also "directly from DAX Device" should be "directly from Device DAX" since that's how the ndctl man page refers to it.
</blockquote></details>

Done.

---

*[src/libpmem/pmem.c, line 687 at r1](https://reviewable.io:443/reviews/pmem/nvml/1675#-KcxRRbkxmPXMYcUnl20-r1-687:-Kd-ctgP9TMHudBybVeB:b-896fix) ([raw file](https://github.com/pmem/nvml/blob/f5d50258e779518f79a803873133a8b999313143/src/libpmem/pmem.c#L687)):*
<details><summary><i>Previously, andyrudoff (Andy Rudoff) wrote…</i></summary><blockquote>

Very picky comment: can we rename the "dax" flag "device_dax" in this function?
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/nvml/1675)*
<!-- Sent from Reviewable.io -->

pbalcer commented on src/common/mmap.h:
```diff
@@ -107,6 +110,39 @@ util_map_hint_align(size_t len, size_t req_align)
+
+int util_range_track(const void *addr, size_t len);
+int util_range_untrack(const void *addr, size_t len);
+map_tracker *util_range_find(const void *addr, size_t len);
```
I think this public interface shouldn't expose the map_tracker structure, but instead return a bool whether or not the provided range is "tracked". That would allow you to hide all of the structures and remove the extern variables.

pbalcer commented on src/common/mmap.c:
```diff
@@ -256,3 +277,187 @@ util_range_none(void *addr, size_t len)
+
+	int ret = 0;
+
+	if (pthread_rwlock_wrlock(&Mmap_list_lock)) {
```
Use the util variant of the locks, it removes the burden of having to handle failures.

pbalcer commented on src/common/mmap.h:
```diff
@@ -107,6 +110,39 @@ util_map_hint_align(size_t len, size_t req_align)
 	return align;
 }
 
+typedef enum map_tracker_flags {
```
We don't use typedefs like this anywhere else in our codebase.

pbalcer commented:
How does this impact running tests on ext4+dax file systems? Is the performance comparable to that of tmpfs with msync? If not, I think we should consider setting the variable PMEM_FS_DIR_FORCE_PMEM to 1 by default in the example testconfig, or at least highly recommending using that option for testing on dax-enabled file systems.

krzycz commented:
I have no perf. data, but I think ext4+dax would be slower. AFAIR, msync() on tmpfs is no-op, while on ext4+dax it actually does something. On the other hand, it should still be faster than ext4 (w/o dax) on SSD/nvme.
I don't think we should force PMEM in our test scripts by default. This may lead to some wrong conclusions and users may be surprised with the time of test execution vs. the performance of the actual apps or even our examples.
Also, keep in mind that ext4+dax sould not be assumed pmem now (from NVML perspective it's a non-pmem file system that requires msync or pmem_msync to flush data to persistence). I think this should be clearly explained in some blog article.

---

Review status: 0 of 88 files reviewed at latest revision, 6 unresolved discussions.

---

*[doc/libpmem.3.md, line 232 at r2](https://reviewable.io:443/reviews/pmem/nvml/1675#-KcxRRbkxmPXMYcUnl1v-r2-232:-Kd4yx55e6XDD-KxbcZK:b-896fix) ([raw file](https://github.com/pmem/nvml/blob/98f453e356b560b5df9b2d45ecfb47af19ac4dcf/doc/libpmem.3.md#L232)):*
<details><summary><i>Previously, pbalcer (Piotr Balcer) wrote…</i></summary><blockquote>

The phrasing on this sentence makes it sound like your PC will explode if you msync a device dax backed mapping.
I'd change it have a meaning more in line with "calling msync on device dax has no effect".
</blockquote></details>

Done.

---

*[src/common/mmap.c, line 444 at r2](https://reviewable.io:443/reviews/pmem/nvml/1675#-KcxRRbkxmPXMYcUnl1x-r2-444:-Kd2Mwf_cFJJUqZcnRGN:bf5w681) ([raw file](https://github.com/pmem/nvml/blob/98f453e356b560b5df9b2d45ecfb47af19ac4dcf/src/common/mmap.c#L444)):*
<details><summary><i>Previously, pbalcer (Piotr Balcer) wrote…</i></summary><blockquote>

Use the util variant of the locks, it removes the burden of having to handle failures.
</blockquote></details>

For some reason there are no util variants for pthread_rwlock_*lock funcs. There is one for pthread_rwlock_unlock only.
Perhaps we could add them, but it's a subject for another patch.

---

*[src/common/mmap.h, line 113 at r2](https://reviewable.io:443/reviews/pmem/nvml/1675#-KcxRRbkxmPXMYcUnl1y-r2-113:-Kd2_s5KUZ7Wr1f64cIw:b-kqwxbu) ([raw file](https://github.com/pmem/nvml/blob/98f453e356b560b5df9b2d45ecfb47af19ac4dcf/src/common/mmap.h#L113)):*
<details><summary><i>Previously, pbalcer (Piotr Balcer) wrote…</i></summary><blockquote>

We don't use typedefs like this anywhere else in our codebase.
</blockquote></details>

What do you mean exactly? What about pmempool common.h for example?

---

*[src/common/mmap.h, line 144 at r2](https://reviewable.io:443/reviews/pmem/nvml/1675#-KcxRRbkxmPXMYcUnl1y-r2-144:-Kd2aBDaz8BNpzMAHz5B:baoy49d) ([raw file](https://github.com/pmem/nvml/blob/98f453e356b560b5df9b2d45ecfb47af19ac4dcf/src/common/mmap.h#L144)):*
<details><summary><i>Previously, pbalcer (Piotr Balcer) wrote…</i></summary><blockquote>

I think this public interface shouldn't expose the map_tracker structure, but instead return a bool whether or not the provided range is "tracked". That would allow you to hide all of the structures and remove the extern variables.
</blockquote></details>

1) map_tracker structure is defined in this header anyway, so it *is* exposed.
2) util_range_find() is used not only to query the region for being pmem or not, but also to find regions to be removed from the list (in which case returning bool is not enough).
Of course, we may define two separate functions for that (the second one would be static), but is this a real issue? This code is a subject for change/refactoring in the near future (1.3 release) anyway. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/nvml/1675#-:-Kd2haAqbwULIFqR3S9m:bgjn0i6)*
<!-- Sent from Reviewable.io -->

pbalcer commented:
Review status: 0 of 88 files reviewed at latest revision, 4 unresolved discussions.

---

*[src/common/mmap.h, line 113 at r2](https://reviewable.io:443/reviews/pmem/nvml/1675#-KcxRRbkxmPXMYcUnl1y-r2-113:-Kd5CCsDLkJt6EHbNjmm:b-yiz9co) ([raw file](https://github.com/pmem/nvml/blob/98f453e356b560b5df9b2d45ecfb47af19ac4dcf/src/common/mmap.h#L113)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary><blockquote>

What do you mean exactly? What about pmempool common.h for example?
</blockquote></details>

We don't usually typedef enums/structs because that pollutes the global namespace. I didn't search the codebase, but you are right, there are two instances of typedef enum/struct... IMHO that's not an argument and we should get rid of those also.

The same issue is below with map_tracker.

---

*[src/common/mmap.h, line 144 at r2](https://reviewable.io:443/reviews/pmem/nvml/1675#-KcxRRbkxmPXMYcUnl1y-r2-144:-Kd5DBg5iZP1lAAKCep8:b-wju9t1) ([raw file](https://github.com/pmem/nvml/blob/98f453e356b560b5df9b2d45ecfb47af19ac4dcf/src/common/mmap.h#L144)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary><blockquote>

1) map_tracker structure is defined in this header anyway, so it *is* exposed.
2) util_range_find() is used not only to query the region for being pmem or not, but also to find regions to be removed from the list (in which case returning bool is not enough).
Of course, we may define two separate functions for that (the second one would be static), but is this a real issue? This code is a subject for change/refactoring in the near future (1.3 release) anyway. 
</blockquote></details>

1. Yes, and my suggestion is to hide it in the .c file, and create equivalent functionality through functions.
2. I'd suggest splitting this into two, one internal with the existing interface and one external with a simpler one.
Yes, it's a real issue because this function right now requires an extern lock(!!) to be held while calling. And I've noticed that temporary solutions have tendency to stick around...

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/nvml/1675)*
<!-- Sent from Reviewable.io -->

pbalcer commented:
I would add a note to the testconfig.sh.example about the fact that "PMEM emulation" doesn't refer to ext4+dax anymore.

---

Review status: 0 of 88 files reviewed at latest revision, 4 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/nvml/1675#-:-Kd5FaS8MRYwETA2m4qC:b-47ozk9)*
<!-- Sent from Reviewable.io -->

krzycz commented:
Review status: 0 of 89 files reviewed at latest revision, 4 unresolved discussions.

---

*[src/common/mmap.h, line 113 at r2](https://reviewable.io:443/reviews/pmem/nvml/1675#-KcxRRbkxmPXMYcUnl1y-r2-113:-Kd5aEgKeTSAisZKhViD:b-896fix) ([raw file](https://github.com/pmem/nvml/blob/98f453e356b560b5df9b2d45ecfb47af19ac4dcf/src/common/mmap.h#L113)):*
<details><summary><i>Previously, pbalcer (Piotr Balcer) wrote…</i></summary><blockquote>

We don't usually typedef enums/structs because that pollutes the global namespace. I didn't search the codebase, but you are right, there are two instances of typedef enum/struct... IMHO that's not an argument and we should get rid of those also.

The same issue is below with map_tracker.
</blockquote></details>

Done.

---

*[src/common/mmap.h, line 144 at r2](https://reviewable.io:443/reviews/pmem/nvml/1675#-KcxRRbkxmPXMYcUnl1y-r2-144:-Kd5aHudUXMbbABU3OKm:b-896fix) ([raw file](https://github.com/pmem/nvml/blob/98f453e356b560b5df9b2d45ecfb47af19ac4dcf/src/common/mmap.h#L144)):*
<details><summary><i>Previously, pbalcer (Piotr Balcer) wrote…</i></summary><blockquote>

1. Yes, and my suggestion is to hide it in the .c file, and create equivalent functionality through functions.
2. I'd suggest splitting this into two, one internal with the existing interface and one external with a simpler one.
Yes, it's a real issue because this function right now requires an extern lock(!!) to be held while calling. And I've noticed that temporary solutions have tendency to stick around...
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/nvml/1675)*
<!-- Sent from Reviewable.io -->

pbalcer commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 0 of 89 files reviewed at latest revision, 4 unresolved discussions, some commit checks broke.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/nvml/1675#-:-Kd5dCuIYOTI_ChlGsZ_:bnfp4nl)*
<!-- Sent from Reviewable.io -->

plebioda commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 38 of 66 files at r1, 67 of 68 files at r2, 5 of 5 files at r3.
Review status: 88 of 89 files reviewed at latest revision, 8 unresolved discussions.

---

*[src/benchmarks/pmembench.cpp, line 1211 at r3](https://reviewable.io:443/reviews/pmem/nvml/1675#-Kd5wsAP0vw1xIvJxyKg:-Kd5wsAQrYUqOQsn32If:b-gt8c0z) ([raw file](https://github.com/pmem/nvml/blob/c33acc80844b187a33feba09c026b77cc79697c5/src/benchmarks/pmembench.cpp#L1211)):*
> ```cpp
> 	}
> 
> 	int ret = 0;
> ```

is this somehow related to pmem_is_pmem changes ?

---

*[src/common/mmap.c, line 356 at r3](https://reviewable.io:443/reviews/pmem/nvml/1675#-Kd6BSMjKouQjCissMZD:-Kd6BSMjKouQjCissMZE:blb57hx) ([raw file](https://github.com/pmem/nvml/blob/c33acc80844b187a33feba09c026b77cc79697c5/src/common/mmap.c#L356)):*
> ```C
> 	int ret = 0;
> 
> 	if (pthread_rwlock_wrlock(&Mmap_list_lock)) {
> ```

util_rwlock_wrlock ? a function which doesn't return if acquiring a lock failed. We use such function for all mutexes and spinlocks.

---

*[src/common/mmap.h, line 113 at r3](https://reviewable.io:443/reviews/pmem/nvml/1675#-Kd5x9c-Vn1lKPTXDVD2:-Kd5x9c-Vn1lKPTXDVD3:b-mqo8pj) ([raw file](https://github.com/pmem/nvml/blob/c33acc80844b187a33feba09c026b77cc79697c5/src/common/mmap.h#L113)):*
> ```objc
> }
> 
> int util_range_track(const void *addr, size_t len);
> ```

just naming... I think register/unregister is better thant track/untrack

---

*[src/libpmem/pmem.c, line 178 at r3](https://reviewable.io:443/reviews/pmem/nvml/1675#-Kd6CLzAqpxuITfB5GIF:-Kd6CLzAqpxuITfB5GIG:b3kre) ([raw file](https://github.com/pmem/nvml/blob/c33acc80844b187a33feba09c026b77cc79697c5/src/libpmem/pmem.c#L178)):*
> ```C
> #include <unistd.h>
> #include <limits.h>
> #include <pthread.h>
> ```

?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/nvml/1675#-:-Kd6Dff7ZwIY8UYUEnc-:bnfp4nl)*
<!-- Sent from Reviewable.io -->

krzycz commented:
Review status: 88 of 89 files reviewed at latest revision, 8 unresolved discussions.

---

*[src/benchmarks/pmembench.cpp, line 1211 at r3](https://reviewable.io:443/reviews/pmem/nvml/1675#-Kd5wsAP0vw1xIvJxyKg:-Kd6KgbfDYFRDDNDOQwd:b9mlqbk) ([raw file](https://github.com/pmem/nvml/blob/c33acc80844b187a33feba09c026b77cc79697c5/src/benchmarks/pmembench.cpp#L1211)):*
<details><summary><i>Previously, plebioda (Pawel Lebioda) wrote…</i></summary><blockquote>

is this somehow related to pmem_is_pmem changes ?
</blockquote></details>

Not really, I added util_mmap_finit() call, but also noticed that argc validation can be done before calloc(). Just a cosmetic change.

---

*[src/common/mmap.c, line 356 at r3](https://reviewable.io:443/reviews/pmem/nvml/1675#-Kd6BSMjKouQjCissMZD:-Kd6L4txsLYiNEWmLWkM:b-3e7f8k) ([raw file](https://github.com/pmem/nvml/blob/c33acc80844b187a33feba09c026b77cc79697c5/src/common/mmap.c#L356)):*
<details><summary><i>Previously, plebioda (Pawel Lebioda) wrote…</i></summary><blockquote>

util_rwlock_wrlock ? a function which doesn't return if acquiring a lock failed. We use such function for all mutexes and spinlocks.
</blockquote></details>

Right, but see my answer to @pbalcer comment.

---

*[src/common/mmap.h, line 113 at r3](https://reviewable.io:443/reviews/pmem/nvml/1675#-Kd5x9c-Vn1lKPTXDVD2:-Kd6KQ7xIJlI3PpiYfA5:b9z6agq) ([raw file](https://github.com/pmem/nvml/blob/c33acc80844b187a33feba09c026b77cc79697c5/src/common/mmap.h#L113)):*
<details><summary><i>Previously, plebioda (Pawel Lebioda) wrote…</i></summary><blockquote>

just naming... I think register/unregister is better thant track/untrack
</blockquote></details>

That is how it was named originally, then I changed it to track/untrack. ;-)

---

*[src/libpmem/pmem.c, line 178 at r3](https://reviewable.io:443/reviews/pmem/nvml/1675#-Kd6CLzAqpxuITfB5GIF:-Kd6JsbDPDky8L9sAYPm:b-896fix) ([raw file](https://github.com/pmem/nvml/blob/c33acc80844b187a33feba09c026b77cc79697c5/src/libpmem/pmem.c#L178)):*
<details><summary><i>Previously, plebioda (Pawel Lebioda) wrote…</i></summary><blockquote>

?
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/nvml/1675)*
<!-- Sent from Reviewable.io -->

krzycz commented:
Review status: 85 of 89 files reviewed at latest revision, 8 unresolved discussions.

---

*[src/common/mmap.h, line 113 at r3](https://reviewable.io:443/reviews/pmem/nvml/1675#-Kd5x9c-Vn1lKPTXDVD2:-Kd6NUbTHd849yQ9DHgE:b-896fix) ([raw file](https://github.com/pmem/nvml/blob/c33acc80844b187a33feba09c026b77cc79697c5/src/common/mmap.h#L113)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary><blockquote>

That is how it was named originally, then I changed it to track/untrack. ;-)
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/nvml/1675)*
<!-- Sent from Reviewable.io -->
